src/lower/mod.rs

```diff
@@ -898,16 +939,73 @@ impl<'r> Lowerer<'r> {
         }
     }
 
+    fn ip_memcmp(&mut self, left: Operand, right: Operand) -> Var {
+        // For an ip addr, we need to be clever, because we can
+        // only compare the bytes that are used the current
+        // variant. Therefore, we read the discriminant of the
+        // left value. If that's 0, we compare 5 bytes,
+        // otherwise, we compare 17 bytes. In other words we
+        // compare 5 + 12 * discriminant bytes.
+
+        // Read discriminant
+        let discriminant = self.new_tmp();
+        self.add(Instruction::Read {
+            to: discriminant.clone(),
+            from: left.clone(),
+            ty: IrType::U8,
+        });
+
+        // Multiply with 12 (so we get 0 or 12)
+        let difference = self.new_tmp();
+        self.add(Instruction::Mul {
+            to: difference.clone(),
+            left: discriminant.into(),
+            right: IrValue::U8(12).into(),
+        });
+
+        // Add 5 (so we get 5 or 17)
+        let size = self.new_tmp();
+        self.add(Instruction::Add {
+            to: size.clone(),
+            left: difference.into(),
+            right: IrValue::U8(5).into(),
+        });
+
+        // Cast the u8 to a usize
+        let size2 = self.new_tmp();
+        self.add(Instruction::Extend {
+            to: size2.clone(),
+            ty: IrType::Pointer,
+            from: size.into(),
+        });
+
+        // Do the actual memcmp
+        let tmp = self.new_tmp();
+        self.add(Instruction::MemCmp {
+            to: tmp.clone(),
+            size: size2.into(),
+            left,
+            right,
+        });
+
+        tmp
+    }
+
     /// Lower a literal
     fn literal(&mut self, lit: &Meta<Literal>) -> Operand {
         match &lit.node {
             Literal::String(_) => todo!(),
             Literal::Asn(n) => IrValue::Asn(*n).into(),
-            Literal::IpAddress(addr) => IrValue::IpAddr(match addr {
-                ast::IpAddress::Ipv4(x) => IpAddr::V4(*x),
-                ast::IpAddress::Ipv6(x) => IpAddr::V6(*x),
-            })
-            .into(),
+            Literal::IpAddress(addr) => {
+                let to = self.new_tmp();
+                const SIZE: usize = std::mem::size_of::<IpAddr>();
+                let x: [u8; SIZE] = unsafe { std::mem::transmute_copy(addr) };
```

Comment:
Is this safe?  The layout of `IpAddr` is not guaranteed, right?

Reply:
It's not safe no. Well, I mean, some parts aren't. Transmuting into some bytes and then transmuting back should be fine, but combined with my own equality checking and all that it becomes unsafe.

Reply:
So that means that I either need to call out for any operation or implement my own ip addr type with stable ABI that hopefully converts back and forth with a no-op. I'd love it if I could just fail to compile if the layout changes lol.